### ISSUE TITLE
Implement Raw Tracepoints and LSM Probes

### ISSUE THREAD

**Issue by willfindlay** [alessandrod/aya#9]
I'm currently working on PR for both of these. Filing this issue to avoid duplication of effort.

**Comment by evilsocket** [comment]
following, i'm waiting for tracepoints support to finally get rid of clang and write ebpf programs in rust

**Comment by willfindlay** [comment]
> following, i'm waiting for tracepoints support to finally get rid of clang and write ebpf programs in rust

To be clear, regular tracepoints are already supported I believe. This is specifically for raw tracepoints, BTF tracepoints, and LSM probes.

**Comment by willfindlay** [comment]
Also, not sure how much time I will have to work on this as I'm currently finishing up a (fairly large) thesis. So if anyone wants to take this over from me in the meantime, feel free. I've got a partial implementation that I can share if needed.

**Comment by evilsocket** [comment]
@willfindlay unless i'm missing something [it doesn't look like regular tracepoints are supported bpf side](https://github.com/alessandrod/aya/tree/main/bpf/aya-bpf/src/programs). Do they work in your fork? I'll gladly help.

**Comment by willfindlay** [comment]
> @willfindlay unless i'm missing something [it doesn't look like regular tracepoints are supported bpf side](https://github.com/alessandrod/aya/tree/main/bpf/aya-bpf/src/programs). Do they work in your fork? I'll gladly help.

Oh you meant aya-bpf, my mistake.

**Comment by evilsocket** [comment]
Yep, I have some eBPF programs that are currently implemented in C and compiled via clang. I'd gladly migrate them to rust and just get rid of the custom compilation step, but they use kprobes and tracepoints so I can't port them yet.

**Comment by tw4452852** [comment]
Hi @evilsocket ,

I just worked out a PR to support this, pls see #29 .

**Comment by evilsocket** [comment]
that's great news @tw4452852! how far is the PR from fully supporting tracepoints and kprobes?

**Comment by vadorovsky** [comment]
@willfindlay I would be interested in LSM program support. If you still have no time to work on that and if you are fine with that, I could try to overtake.

**Comment by willfindlay** [comment]
> @willfindlay I would be interested in LSM program support. If you still have no time to work on that and if you are fine with that, I could try to overtake.

Sure feel free to take over, as I'm still quite busy. I didn't end up getting very far with this, but let me know if there's anything I can help with.

**Comment by vadorovsky** [comment]
I still didn't get it fully done, but I will share what I have for now.

I pushed aya changes (based on @willfindlay's work) to my fork in this branch:
https://github.com/vadorovsky/aya/tree/lsm

And here I have an example of program I'm trying to load:
https://github.com/vadorovsky/aya-example-lsm

It still doesn't work though. I'm getting the following error:

```
❯ sudo ./target/debug/aya-example-lsm 
[sudo] password for root: 
error: the BPF_PROG_LOAD syscall failed. Verifier output: func#0 @0
verification time 5 usec
stack depth 0
processed 0 insns (limit 1000000) max_states_per_insn 0 total_states 0 peak_states 0 mark_read 0
: Invalid argument (os error 22)
```

It might be related to the fact that there is an additional parameter needed for loadiong (`BPF_PROG_LOAD`) LSM programs - `attach_btf_id`.

I'm retrieving a value for that parameter here:
https://github.com/vadorovsky/aya/commit/053a2b261ff3bc911e31d2ccd5f335ed70d0eae6#diff-680ebb3c74bd2296b1ccdf57ed77ee96ea9cd67896d72815313fb040d97fa313R227-R252
https://github.com/vadorovsky/aya/commit/053a2b261ff3bc911e31d2ccd5f335ed70d0eae6#diff-89a5a1aae78a85cd7324868565b540c2e11fb332e25dab9a4d4f354a0e24fd1bR44-R53

It might be some mistake in that logic. I made it way simplier than upstream libbpf. When I tried debugging libbpf, I noticed that its `btf__find_by_name_kind` [function](https://github.com/torvalds/linux/blob/d4d016caa4b85b9aa98d7ec8c84e928621a614bc/tools/lib/bpf/btf.c#L692-L712) is always called with parameters like:

```
* thread #1, name = 'lockcd', stop reason = breakpoint 1.1
    frame #0: 0x0000555555663d80 lockcd`btf__find_by_name_kind(btf=0x0000555555815d50, type_name="bpf_lsm_task_alloc", kind=12) at btf.c:698:11
```

```
* thread #1, name = 'lockcd', stop reason = breakpoint 1.1
    frame #0: 0x0000555555663d80 lockcd`btf__find_by_name_kind(btf=0x0000555555815d50, type_name="bpf_lsm_syslog", kind=12) at btf.c:698:11
```

so with `type_name="bpf_lsm_{function_name}, kind=BTF_TYPE_FUNC`.

I will try to dig into it this week.

**Comment by willfindlay** [comment]
@vadorovsky I believe the issue is you need to load the program using its BTF info. That's where I ended up getting too busy to continue unfortunately. My plan was to reverse engineer the way libbpf handles program loading with BTF. That may be a good place to start.

**Comment by vadorovsky** [comment]
@willfindlay Thanks for the hint. I completelly missed the part of libbpf/btf.c which gets BTF fd by id. I know what to look for now, thanks!

**Comment by vadorovsky** [comment]
Good news - I got a simple LSM BPF program working. I will make a draft PR soon.

Another good news.
@willfindlay Using program's BTF info was not needed finally. At least not for LSM programs. Debugging plain libbpf showed me the following attrs for LSM programs:

```
(const bpf_prog_load_params) $8 = {
  prog_type = BPF_PROG_TYPE_LSM
  expected_attach_type = BPF_LSM_MAC
  name = 0x0000555555b5dd10 "mount_audit"
  insns = 0x0000555555b600c0
  insn_cnt = 350
  license = 0x0000555555b5d780 "GPL"
  kern_version = 331275
  attach_prog_fd = 0
  attach_btf_obj_fd = 0
  attach_btf_id = 31801
  prog_ifindex = 0
  prog_btf_fd = 3
  prog_flags = 0
  func_info_rec_size = 8
  func_info = 0x0000555555b611c0
  func_info_cnt = 2
  line_info_rec_size = 16
  line_info = 0x0000555555b61690
  line_info_cnt = 69
  log_level = 0
  log_buf = 0x0000000000000000
  log_buf_sz = 0
}
```

These are parameters it used for loading https://github.com/rancher-sandbox/lockc/blob/main/lockc/src/bpf/lockc.bpf.c#L298-L413

The thing which made me stuck before was... lack of `expected_attach_type` in `load_attrs`. Setting `expected_attach_type` to `BPF_LSM_MAC` made it working.

**Comment by vadorovsky** [comment]
Check the last push, the previous one still didn't work (It also had `BPF_BTF_GET_NEXT_ID` and `BPF_BTF_GET_FD_BY_ID` syscalls added, which came out to be unnecessary). :)
